nixos/modules/services/matrix/synapse/workers.nix

```diff
@@ -0,0 +1,198 @@
+              description = "The type of worker application";
+              default = "synapse.app.generic_worker";
+            };
+            worker_replication_host = mkOption {
+              type = types.str;
+              default = cfg.workers.mainReplicationHost;
+              defaultText = lib.literalExpression "config.services.matrix-synapse.workers.mainReplicationHost";
+              description = "The replication listeners IP on the main synapse process";
+            };
+            worker_replication_http_port = mkOption {
+              type = types.port;
+              default = cfg.workers.mainReplicationPort;
+              defaultText = literalExpression "config.services.matrix-synapse.workers.mainReplicationPort";
+              description = "The replication listeners port on the main synapse process";
+            };
+            worker_listeners = mkOption {
+              type = types.listOf workerListenerType;
+              default = [ ];
+              description = "Listener config for the worker, similar to the main synapse listener config";
+            };
+          };
+        };
+        default = { };
+        description = "A freeform option of the workers settings, see upstream synapse documentation for options";
+      };
+    });
+
+    workerListenerType = types.submodule {
+      options = {
+        type = mkOption {
+          type = types.enum [ "http" "metrics" ];
+          description = "The type of the listener";
+          default = "http";
+        };
+        port = mkOption {
+          type = types.port;
+          description = "The TCP port to bind to";
+        };
+        bind_addresses = mkOption {
+          type = with types; listOf str;
+          description = "A list of local addresses to listen on";
+          default = [ "127.0.0.1" ];
+        };
+        tls = mkOption {
+          type = types.bool;
+          description = ''
+            Whether to enable TLS for this listener.
+            Will use the TLS key/cert specified in tls_private_key_path / tls_certificate_path.
+          '';
+          default = false;
+        };
+        x_forwarded = mkOption {
+          type = types.bool;
+          description = ''
+            Whether to use the X-Forwarded-For HTTP header as the client IP.
+            This option is only valid for an 'http' listener.
+            It is useful when Synapse is running behind a reverse-proxy.
+          '';
+          default = true;
+        };
+        resources = mkOption {
+          type = types.listOf (types.submodule {
+            options = {
+              names = mkOption {
+                type = types.listOf (types.enum [
+                  "client"
+                  "consent"
+                  "federation"
+                  "keys"
+                  "media"
+                  "metrics"
+                  "openid"
+                  "replication"
+                  "static"
+                  "webclient" # TODO: wtf?
```

Comment:
The `"health"` resource (see https://matrix-org.github.io/synapse/latest/usage/configuration/config_documentation.html#listeners) is missing. It's very useful for workers like a federation sender which don't have any other resources configured.
Also I'm not sure where the `"webclient"` resource is from as that doesn't seem to be documented upstream.

Reply:
It's very old and is what is served on `/client` I believe.  But it should be removed yep. Thanks for pointing out health